crates/kubelet/src/resources/device_plugin_manager/manager.rs

```diff
@@ -0,0 +1,844 @@
+    }
+
+    /// Adds the plugin to our HashMap
+    async fn add_plugin(&self, plugin_connection: Arc<PluginConnection>, resource_name: &str) {
+        let mut lock = self.plugins.write().await;
+        lock.insert(resource_name.to_string(), plugin_connection);
+    }
+
+    /// Validates the given plugin info gathered from a discovered plugin, returning an error with
+    /// additional information if it is not valid. This will validate 3 specific things (should
+    /// answer YES to all of these):
+    /// 1. Does this manager support the device plugin version? Currently only accepting `API_VERSION`.
+    ///    TODO: determine whether can support all versions prior to current `API_VERSION`.
+    /// 2. Does the plugin have a valid extended resource name?
+    /// 3. Is the plugin name available? 2a. If the name is already registered, is the plugin_connection the
+    ///    exact same? If it is, we allow it to reregister
+    pub(crate) async fn validate(
+        &self,
+        register_request: &RegisterRequest,
+    ) -> Result<(), tonic::Status> {
+        trace!(
+            "Starting validation for plugin {:?} discovered at path {}",
+            register_request.resource_name,
+            register_request.endpoint
+        );
+
+        // Validate that version matches the Device Plugin API version
+        if register_request.version != API_VERSION {
+            return Err(tonic::Status::new(
+                tonic::Code::Unimplemented,
+                format!(
+                    "kubelet doesn't support version Device Plugin {}",
+                    API_VERSION
+                ),
+            ));
+        };
+
+        // Validate that plugin has proper extended resource name
+        if !util::is_extended_resource_name(&register_request.resource_name) {
+            return Err(tonic::Status::new(
+                tonic::Code::Unimplemented,
+                format!(
+                    "resource name {} is not properly formatted. See https://github.com/kubernetes/community/blob/master/contributors/design-proposals/scheduling/resources.md#resource-types", register_request.resource_name)
+            ));
+        }
+
+        self.validate_is_unique(register_request).await?;
+
+        Ok(())
+    }
+
+    /// Validates if the plugin is unique (meaning it doesn't exist in the `plugins` map).
+    /// If there is an active plugin registered with this name, returns error.
+    /// TODO: Might be best to always accept: https://sourcegraph.com/github.com/kubernetes/kubernetes@9d6e5049bb719abf41b69c91437d25e273829746/-/blob/pkg/kubelet/cm/devicemanager/manager.go?subtree=true#L439
+    async fn validate_is_unique(
+        &self,
+        register_request: &RegisterRequest,
+    ) -> Result<(), tonic::Status> {
+        let plugins = self.plugins.read().await;
+
+        if let Some(_previous_plugin_entry) = plugins.get(&register_request.resource_name) {
+            // TODO: check if plugin is active
```

Comment:
I think we can remove the `validate_is_unique` function as Kubernetes' current implementation replaces existing device plugins with the same resource name: https://sourcegraph.com/github.com/kubernetes/kubernetes@9d6e5049bb719abf41b69c91437d25e273829746/-/blob/pkg/kubelet/cm/devicemanager/manager.go?subtree=true#L439. They have a todo on themselves to do more explicit checking as to whether the plugin is alive, though.

Reply:
Ok, can you please drop the link to the existing TODO in k8s so that people can follow along if they come along later?

Reply:
even though it doesnt exactly match the k8s implementation, i think we should stick with the current implementation of `validate_is_unique` not allowing a plugin to register if there already is one with that name already registered. K8s allows the newest plugin to replace the old but (according to their TODO https://sourcegraph.com/github.com/kubernetes/kubernetes@9d6e5049bb719abf41b69c91437d25e273829746/-/blob/pkg/kubelet/cm/devicemanager/manager.go?subtree=true#L439) is striving for the behavior we have implemented of only replacing the plugin if it is no longer alive. In the krustlet device manager, the old plugin is still alive by the fact that the connection to it is still maintained, otherwise it wouldve been removed.